components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/ServiceDiscoverer.java

```diff
@@ -0,0 +1,151 @@
+     *
+     * @param namespace     Namespace of the expected endpoints
+     * @return List of Endpoints with the specified namespace
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(String namespace) throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints filtered by criteria.
+     *
+     * @param criteria    A set of criteria which all the endpoints must belong to
+     * @return List of Endpoints with the specified criteria
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(HashMap<String, String> criteria) throws ServiceDiscoveryException;
+
+    /**
+     * Gives a list of endpoints filtered by both namespace and criteria.
+     *
+     * @param namespace   Namespace of the expected endpoints
+     * @param criteria    A set of criteria which all the endpoints must belong to
+     * @return List of Endpoints with the specified namespace and criteria
+     * @throws ServiceDiscoveryException if an error occurs while listing
+     */
+    public abstract List<Endpoint> listServices(String namespace, HashMap<String, String> criteria)
+            throws ServiceDiscoveryException;
+
+    /**
+     * Builds an Endpoint
+     *
+     * @param id                Temporary id to be used by the UI
+     * @param name              Name of the service
+     * @param endpointConfig    Json string containing endpoint URL, namespace, criteria
+     * @param maxTps            MaxTps
+     * @param type              Application level protocol (eg. http/https)
+     * @param endpointSecurity  Json string on endpoint security necessarily including "enabled" boolean key
+     * @param applicableLevel   Whether applicable level is global or production only
+     * @return {@link org.wso2.carbon.apimgt.core.models.Endpoint} object
+     */
+    protected Endpoint buildEndpoint(String id, String name, String endpointConfig, Long maxTps,
+                                    String type, String endpointSecurity, String applicableLevel) {
+        Endpoint.Builder endpointBuilder = new Endpoint.Builder();
+        endpointBuilder.id(id);
+        endpointBuilder.name(name);
+        endpointBuilder.endpointConfig(endpointConfig);
+        endpointBuilder.maxTps(maxTps);
+        endpointBuilder.type(type);
+        endpointBuilder.security(endpointSecurity);
+        endpointBuilder.applicableLevel(applicableLevel);
+
+        serviceEndpointIndex++;
+        return endpointBuilder.build();
+    }
+
+    /**
+     * Gives the namespace to filter by
+     *
+     * @return namespace
+     */
+    public String getNamespaceFilter() {
+        return namespaceFilter;
+    }
+
+    /**
+     * Gives the criteria to filter by
+     *
+     * @return criteria map
+     */
+    public HashMap<String, String> getCriteriaFilter() {
```

Comment:
return generic map instead of HashMap

Reply:
There was no special reason to have HashMap in those places. I'll change all of them (at declarations) from HashMap to Map.